src/ObjectUtils/PdbFile.cpp

```diff
@@ -0,0 +1,149 @@
+
+// Codeview debug records from a PDB file can be accessed through llvm using a visitor
+// interface (using llvm::codeview::CVSymbolVisitor::visitSymbolStream). This can be
+// customized by implementing one's own visitor class, which we do here to fill out
+// all symbol info required for functions.
+class SymbolInfoVisitor : public llvm::codeview::SymbolVisitorCallbacks {
+ public:
+  SymbolInfoVisitor(ModuleSymbols* module_symbols) : module_symbols_(module_symbols) {}
+
+  // This is the only record type (ProcSym) we are interested in, so we only override this
+  // method. Other records will simply return llvm::Error::success without any work done.
+  llvm::Error visitKnownRecord(llvm::codeview::CVSymbol& record,
+                               llvm::codeview::ProcSym& proc) override {
+    SymbolInfo symbol_info;
+    symbol_info.set_name(proc.Name.str());
+    symbol_info.set_demangled_name(llvm::demangle(proc.Name.str()));
+    // This address is a relative virtual address (RVA).
+    symbol_info.set_address(proc.CodeOffset);
+    symbol_info.set_size(proc.CodeSize);
+    *(module_symbols_->add_symbol_infos()) = std::move(symbol_info);
+
+    return llvm::Error::success();
+  }
+
+ private:
+  ModuleSymbols* module_symbols_;
+};
+
+class PdbFileImpl : public PdbFile {
+ public:
+  PdbFileImpl(std::filesystem::path file_path, std::unique_ptr<llvm::pdb::IPDBSession> session);
+
+  [[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> LoadDebugSymbols() override;
+  [[nodiscard]] const std::filesystem::path& GetFilePath() const override { return file_path_; }
+
+  [[nodiscard]] const std::array<uint8_t, 16> GetGuid() const override {
+    constexpr int kGuidSize = 16;
+    std::array<uint8_t, kGuidSize> result;
+    auto global_scope = session_->getGlobalScope();
+    const llvm::codeview::GUID& guid = global_scope->getGuid();
+    std::copy(&(guid.Guid[0]), &(guid.Guid[kGuidSize - 1]), &result[0]);
+    return result;
+  }
+
+  [[nodiscard]] uint32_t GetAge() const override { return session_->getGlobalScope()->getAge(); }
+
+ private:
+  std::filesystem::path file_path_;
+  std::unique_ptr<llvm::pdb::IPDBSession> session_;
+};
+
+PdbFileImpl::PdbFileImpl(std::filesystem::path file_path,
+                         std::unique_ptr<llvm::pdb::IPDBSession> session)
+    : file_path_(std::move(file_path)), session_(std::move(session)) {}
+
+[[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> PdbFileImpl::LoadDebugSymbols() {
+  ModuleSymbols module_symbols;
+  module_symbols.set_symbols_file_path(file_path_.string());
+
+  llvm::pdb::NativeSession* native_session = static_cast<llvm::pdb::NativeSession*>(session_.get());
+  llvm::pdb::PDBFile& pdb_file = native_session->getPDBFile();
+
+  if (!pdb_file.hasPDBDbiStream()) {
+    return ErrorMessage("PDB file does not have a DBI stream.");
+  }
+
+  auto& debug_info_stream = pdb_file.getPDBDbiStream();
+  const auto& modules = debug_info_stream->modules();
```

Comment:
nit: maybe mention types explicitly in this case - it is not obvious what modules() returns for example.

Reply:
Done.